### DISCUSSION TITLE
Datasources in Perses

### DISCUSSION THREAD

**Discussion by LukeTillman** [perses/perses#404] [Design Discussions]
After some conversation with @Nexucis on Matrix/Zoom, I wanted to write up a proposal for how Datasources will work in Perses.

## What is a Datasource?

Conceptually, a Datasource in Perses is just another system that we communicate with to retrieve data. For example, Perses might communicate with Prometheus, Graphite, or MySQL to retrieve data for display on a graph or to retrieve data that's meant to be displayed as template variable value options. A user might have multiple of the same **kind** of system that they want to communicate with. For example, a user might have three different Prometheus instances that they want to retrieve data from, each with different configuration like the base URL for HTTP endpoints.

In Perses we want to be able to display a list of Datasources to the user _for a particular kind_ so that they can:

- select a Datasource to be used for a graph query
- select a Datasource to be used for a template variable options query
- select a Datasource to be used as the default for a Dashboard
- select a Datasource to be used as the default for a Project
- select a Datasource to be used as the default for Perses (i.e. globally)

It's important to note the "for a particular kind" for two reasons. First, only one kind of Datasource will make sense for a given query. For example, you would want a `PrometheusDatasource` in order to make a `PrometheusLabelValues` query. Second, we want the defaults for Dashboards/Projects/Perses to be the default per kind of Datasource to allow users to take advantage of fallback behavior in their queries (and not have to specify an exact Datasource for every single query in a Dashboard definition).

## Defining a Datasource

A Datasource definition will consist of the following common properties:

```ts
interface DatasourceDefinition<Options> {
  kind: string;
  display: {
    name: string;
  };
  options: Options;
}
```

Like other definitions that are controlled by plugins, `Options` will be specific to the kind of Datasource being defined, so for example a really simple `PrometheusDatasource` might have options that look like:

```ts
interface PrometheusDatasourceOptions {
  baseUrl: string;
}
```

Datasources can be defined at the Dashboard level. So the Perses Dashboard spec will be extended to include:

```ts
interface DashboardSpec {
  // ... existing dashboard spec ...
  datasources: Record<string, DatasourceDefinition>;
}
```

Each Datasource has a unique string key which will be used to uniquely identify it in selectors (more on this below). So a sample spec might look like this:

```jsonc
{
  // ... rest of spec ...
  "datasources": {
    "PromDemo": {
      "kind": "PrometheusDatasource",
      "display": {
        "name": "Prometheus Demo Server"
      },
      "options": {
        "baseUrl": "https://sample.com"
      }
    },
    "GraphiteDemo": {
      "kind": "GraphiteDatasource",
      "display": {
        "name": "Graphite Demo Server"
      },
      "options": {}
    }
  }
}
```

Datasources can also be defined at the Project level, allowing all the dashboards in a Project to use it. Since in some deployment scenarios on Kubernetes a "Project" will really just be an alias for a K8s namespace, we don't want to extend the current `Project` resource to define datasources there since that would imply that a User would need permission to edit a K8s namespace in order to edit the Datasources.

> Note: for more details on the Project/K8s namespace deployment scenarios, see [TODO: @Nexucis to provide some docs explaining what we discussed].

Instead, we'll introduce a new resource type that lives inside of the Project (namespace) that will look like this initially:

```ts
interface ProjectConfigResource {
  kind: "ProjectConfig";
  metadata: Metadata;
  spec: ProjectConfigSpec;
}

interface ProjectConfigSpec {
  datasources: Record<string, DatasourceDefinition>;
}
```

Just like dashboards, `datasources` will have a unique key for identifying the Datasource.

Finally, Datasources can be defined at the global/application level. In K8s deployment scenarios, we envision the Perses application being deployed to its own namespace and any global datasources could be defined in the application configuration that lives in that namespace with the app. Since we don't currently have a configuration resource like that, we'll introduce a new one that looks something like this:

```ts
interface PersesConfigResource {
  kind: "PersesConfig";
  metadata: Metadata;
  spec: PersesConfigSpec;
}

interface PersesConfigSpec {
  datasources: Record<string, DatasourceDefinition>;
}
```

Again, we will enforce a unique key in `datasources`. Any Datasources defined at the global level would be available to Dashboards across all Projects.

## Selecting / Referencing a Datasource

For queries that need a Datasource, we need some way for the user to specify _which_ Datasource to use for a given query. For this purpose we'll introduce a selector that looks something like this:

```ts
interface DatasourceSelector {
  kind: string;
  key?: string;
}
```

This means that when a user selects a Datasource they are _always_ selecting a kind and _optionally_ selecting a specific Datasource with the specified key. If the `key` is omitted, that effectively means "use the default datasource for this kind".

> **TODO / For Discussion**: since keys are already unique, do we really even need `kind` as part of the selector? At runtime, the Prometheus plugin will know what `kind` it needs to ask for, so keeping `kind` in the selector is really only useful if we want to do some validation before runtime that a selector can actually be resolved to the appropriate `kind`. If we don't want to keep `kind`, then a selector can really just be the `string` unique key.

So for example, to use the Prometheus Datasource with the key `PromDemo`, you would use a selector like this:

```jsonc
{
  "kind": "PrometheusDatasource",
  "key": "PromDemo"
}
```

Or to use the default Prometheus Datasource, you would use this:

```jsonc
{
  "kind": "PrometheusDatasource"
}
```

More on how these selectors are resolved in a moment, but first let's talk about defaults.

## Default Datasources

A user should be able to specify a default Datasource for a particular kind of Datasource. We want them to be able to specify this at the Dashboard, Project, and Global levels. That means further extending some of the specs above to look like this:

```ts
// Required<> here just means all properties are required (i.e. you must
// specify both a key and a kind)
type KeyDatasourceSelector = Required<DatasourceSelector>;

// At the Dashboard level
interface DashboardSpec {
  // ... existing dashboard spec ...
  datasources: Record<string, DatasourceDefinition>;
  defaultDatasources: Array<KeyDatasourceSelector>;
}

// At the Project level
interface ProjectConfigSpec {
  datasources: Record<string, DatasourceDefinition>;
  defaultDatasources: Array<KeyDatasourceSelector>;
}

// At the Global level
interface PersesConfigSpec {
  datasources: Record<string, DatasourceDefinition>;
  defaultDatasources: Array<KeyDatasourceSelector>;
}
```

There are two things to note here about the `defaultDatasources` array. First, we will enforce that a given `kind` only appears once, meaning you can only specify one default per kind. Second, this is an array of `KeyDatasourceSelector`, meaning you can only select a specifc Datasource (i.e. with a specific key) as a default for a particular kind.

> **TODO / For Discussion**: An alternative syntax for `defaultDatasources` might just be `Record<string, string>` where the key in the `Record` is the Datasource `kind` and the value in the `Record` is the Datasource's unique key. I went with selectors instead but the other way definitely might be better?

So in practice, you might end up with something like this:

```jsonc
{
  // ... rest of spec with datasources ...
  "defaultDatasources": [
    {
      "kind": "PrometheusDatasource",
      "key": "PromDemo"
    },
    {
      "kind": "GraphiteDatasource",
      "key": "GraphiteDemo"
    }
  }
}
```

Which leads to the question, how are these selectors resolved?

## Resolving Datasource selectors

The resources in Perses currently have the following hierarchy:

- A Dashboard belongs to...
- A Project which belongs to...
- A Perses application

When resolving Datasources, we'll fall back in the same order, starting from where the selector is being used. To "fall back" here basically means to move to the next resource in the hierarchy (i.e. from Dashboard -> Project -> Application).

So for a given `$SELECTOR` in `$CURRENT_SPEC`, the basic algorithm for resolving the selector is:

1. If `$SELECTOR` does not have a `key`, try to find a default for `$SELECTOR.kind` in the `$CURRENT_SPEC.defaultDatasources`. If a match is found, replace `$SELECTOR` with that match. If it already has a `key`, go to next step.
2. If `$SELECTOR` has a `key`, try to find a match on `key` and `kind` in `$CURRENT_SPEC.datasources`. If no match is found or `$SELECTOR` does not have a `key`, go to next step.`
3. Fall back by replacing `$CURRENT_SPEC` with the next resource in the hierarchy. If there is not another resource in the hierarchy to fall back to, error out.
4. Repeat steps 1-3.

This fallback behavior has a few benefits:

1. Resources lower in the resource hierarchy can potentially override ones higher up. (e.g. a Datasource definition in a Dashboard could override a global one in the Perses config).
2. Datasource definitions can move up in the hierarchy without breaking selectors. (e.g a Datasource definition could start in a Dashboard, but move to the Project/Global level without breaking selectors in that Dashboard).

## What is a Datasource Plugin in the UI?

TODO: Once we are agreed on the spec/definitions/etc., I'll try to flesh this out a bit.

**Comment by LukeTillman** [comment]
There's still an open question I think about the scenario where we want the backend to act as a proxy for a Datasource. This could be for security reasons (i.e. don't want to expose Prometheus directly to the browser) or because for certain Datasources, it wouldn't ever make sense to have the UI talking directly to it (e.g. MySQL). I think we should consider maybe making that a separate "thing" in Perses (i.e. resource?) and maybe talk through what the backend needs to setup the appropriate endpoints for the UI to talk to, etc. 

But that could happen in a separate discussion. I'm really thinking of a Datasource as a plugin for the UI for talking to other systems right now and that other system could be the Perses backend proxying once we flesh out the other use case.

**Reply by Nexucis** [reply]
From my point of view, that's a difficult topic and I don't think we can move forward without taking time to consider it.

Yes the backend will need to act as a proxy it's mandatory to have this feature. For example, without this, we can't use Perses in Amadeus.

Another example is Perses deployed in k8s, it will be hard for the UI to be able to talk directly to the datasources.

Once we acknowledge the BE will be a datasource proxy, it implies that not everything in a datasource can be totally free to define. The BE will need to know for example the HTTP configuration. Which implies this HTTP config won't be something that can be changed in the datasource plugin. At most the Datasource definition will just reference the HTTP config that is known and defined in the BE

**Comment by sjcobb** [comment]
This all sounds great, one note, assuming we will eventually have a UI for adding datasources [similar to Grafana](https://grafana.com/docs/grafana/latest/datasources/add-a-data-source/), I could see a few potential new properties within display such as info (for description text) and logo (could also show beside name in datasource selector).

**Reply by Nexucis** [reply]
yeah I think having a page to create / update datasource will be required.

**Comment by AntoineThebaud** [comment]
Neat! These scoped defaults are a great idea.

Just about this fallback based on defaults & the scope 'widening' if no match is found: we could end up in a situation where a lower-level-scoped datasource 'hides' an upper-level-scoped datasource, although you'd like to use both in your dashboard for different graphs. E.g:
- the project-default <kind> datasource & the dashboard-default <kind> datasource have the same key
- a new project-scoped datasource is added after your already finished creating a dashboard. This new datasource has the same kind + key as the dashboard-scoped datasource you defined, but you actually would like to edit your dashboard to use this new project-scoped datasource for some graphs.

I'm not sure that's a real issue / a use case that makes sense, because you could just rename your dashboard-scoped datasources to avoid such conflicts and that's it. But in the second example where you were the first one to use a given name, and have to change it because the same name was later used for a new project-scoped datasource, it can be annoying.

A way to deal with this would basically consist in doing something like:
```typescript
interface DatasourceSelector {
  kind: string;
  key?: string;
  level?: string;
}
```
with level being like:
- level=_empty_ -> search datasource by going through Dashboard -> Project -> Application
- level="Dashboard" -> search datasource by going through Dashboard -> Project -> Application (useless duplicate maybe?)
- level="Project" -> search datasource by going through Project -> Application
- level="Application" -> search datasource by going through Application

Level would probably be empty in most cases, thus that's no additional "weight" for standard users. On the other hand people that would like to use a 'hidden' datasource could specify this attribute in the given datasource selector, instead of changing the key attribute in the datasource + in all the selectors refering to it.

But again, maybe I'm going to far here & imagining a theoretical situation that will never happen in practice :D

**Reply by Nexucis** [reply]
As a side note here, I think it will be interesting to display in the UI where the datasource chosen is coming from. Does it come from the dashboard itself, from the project or from the global definition ?

**Reply by LukeTillman** [reply]
I was thinking of the "hiding" aspect as a feature that someone might want to take advantage of. So, for example, if I want to override how a project-level datasource works (and I don't have permissions to edit the project), I could just do it for my dashboard. Then if my overrides make sense, I might try and get them promoted as the defaults for the project. That's why I purposefully omitted the "level" selector in the proposal.

**Reply by AntoineThebaud** [reply]
Yes as I said I think this "hiding" aspect is a good idea, here I just would like to provide an optional parameter that wont prevent this, it would just allow to bypass this behavior if ever you have to (see the use cases/examples I had in mind above).
In terms of UI design, I imagine this as a tuning option that shouldn't appear by default (something accessible via right click, or "..." button, etc)

**Comment by AntoineThebaud** [comment]
@LukeTillman just to remove a possible ambiguity here: in your design, for a given panel, would we be able to provide a datasource selector at panel level or query level?

**Reply by Nexucis** [reply]
I think a question related to this point is also: do we want to select multiple datasources for the same panel ? So we can draw lines with data coming from different datasources.

**Reply by LukeTillman** [reply]
I was thinking that it would be at the query level because, for example, you might want to graph data from mixed data sources in the same panel.

**Reply by Nexucis** [reply]
I have a bad feeling about mixing datasources in the same panel. Because it means you could have a SQL and a Prometheus datasource in the same panels, with different scale. With let's say 2 queries it might work because you could have two Y axes (one on the left and another one on the right). But then with 3 queries, you might have an issue with the scale.

And even, I'm really wondering if we have a requirement to mixed datasources in the same dashboard.

From my point of view, I would allow the selection of the datasource at panel level like Grafana for simplicity.

**Reply by LukeTillman** [reply]
Yeah, I agree it's probably not ideal (and maybe not a common use case?), but I meant to call out "cause Grafana lets you do it" as a reason for doing it this way too. They do let you mix datasources in the same panel (there's an option for "Mixed" at the panel level). I guess my question would be, do we need to support that Grafana feature?

If not, cool with me if we do it at the panel level. If we do, I think we should do it at the query level to avoid the Grafana situation of having datasource both for the panel and query in the JSON.

**Comment by Nexucis** [comment]
Thank you @LukeTillman for taking the time to tackle this pain topic.

Overall it looks ok specially regarding the changes on the dashboard spec.

I have more concerns about introducing a new config object in the project and then at the root level containing a list of Datasource.

The issue I have with that is based on how the API is designed. Today the API is following the REST principals which means:

For each resources provided, you have the following HTTP endpoints:
* `GET /<resources_type>` which returns a list of the resources stored
* `POST /<resources_type>` which allows you to create an unique resource
* `PUT /<resources_type>/<resource_name>` which allows you to update an unique resource if it exists
* `DELETE /<resources_type><resource_name>` which allows you to delete it.

And you don’t have anything else to interact with a resource. Looking like that, we could think it’s a pretty basic usage and basic implementation of an API. But I can assure you that following strictly these principals help a lot to keep the backend clean and maintainable, because then it means every resources are managed in the same “way”.

Why on earth I am talking about that ? That’s because if you want to introduce a new object like ProjectConfigSpec, you have to think you will be able to create multiple object ProjectConfigSpec and not a single one.

Saying differently, it means you will be able to create multiple list of datasources per projects or at the root level. So from my point of view, the point of storing a list of datasource in a unique object doesn’t follow the REST principales or cancel them.

I know it’s convenient on the FE point of view, specially regarding the default datasource. But I can assure you that breaking the REST principals is a really bad idea and we will regret it later when adding more features around the datasources.

My counter proposal is to keep one object per datasource, but we will provide more query parameters to be able to filter the datasources. For example we could imagine we have a way to get the list of the Prometheus datasources in the project perses: `GET /projects/perses/datasources?kind=Prometheus`

We could imagine the same thing for the default datasources. For example `GET /projects/perses/datasources?default=true` will return the list of all default datasources stored in the project Perses.

And then if we combine this two query parameter `GET /projects/perses/datasources?default=true&kind=Prometheus` then it should return an array with a single element or an empty array if the default Prometheus datasource hasn't beed defined.

With this kind of filter, when the FE will look at the datasource selector in the dashboard spec, I am pretty sure you will have the required information to perform a query that will return the datasource needed.

**Reply by LukeTillman** [reply]
Interesting, I think I'm following what you're saying. I'm not opposed to keeping them as a separate resource, but I have a few questions:

1. If they're separate resources, do we keep the idea of having three "levels" (i.e. at the dashboard, project, and global levels)?
2. If we keep the three levels, how do I associate a resource with a particular level (i.e. how do I say "this datasource is only for this dashboard" or "this datasource belongs to this project" or "this datasource is global")? It feels like one thing we should be trying to avoid is having different resource kinds for each level since they're all basically the same resource, just with different scope.
3. How do I specify the default for a given level (i.e. how do I say "this datasource is the default Prometheus datasource for this project")? 

I think if we can come up with a few example specs (or maybe just some pseudo-code type definitions of what the resource might look like), this will all make a lot more sense. Then we can probably map out what APIs we need (e.g. how does the UI query for datasources at each level).

**Reply by Nexucis** [reply]
1. Yes totally, it doesn't change anything on this vision
2. Today with the Configuration object, you will have a different kind of object for global config and project config. So on this side, it doesn't change as well. You will have a different kind of resources depending of the scope. Because that's definitely the issue, with rest API when you have a different scope, I can assure you the simplest way is to have different kind of resource even if at the end they can be considered as the same.

And on the frontend, it doesn't change so much things, it will just change the http API call basically which is fine because it will definitely depends on which level you are looking for to get your datasource: locally in your dashboard, in your project, globally.

3. With a simple boolean in the spec definition of the datasource. like `default: true`

**Reply by Nexucis** [reply]
and if you need to simplify the datasource management in the frontend side, to reduce the number of object, you could do something like that: 


```typescript
type GlobalDatasource interface {
   kind: "GlobalDatasource"
   metadata: Metadata
   spec: DatasourceSpec
}

type Datasource interface {
   kind: "Datasource"
   metadata: ProjectMetadata
   spec: DatasourceSpec
}
``` 

or

```typescript
type Datasource interface {
   kind: "Datasource" | "GlobalDatasource"
   metadata: Metadata | ProjectMetatada
   spec: DatasourceSpec
}
```

**Comment by Nexucis** [comment]
Also I think it will be great to replace the `option` field by `spec` so we have a single attribute that is expressing it will contain the configuration of an object or a sub part of an object.

**Reply by LukeTillman** [reply]
Oh man, I was wondering if we'd talk about this design choice at some point! 😂 

We've followed the `options` convention for plugins to try and keep all the "plugin specific" options together on a single type that's a nested field. That was an intentional choice to try and prevent conflicts between common options and plugin-specific ones, as well as (to use a little object-oriented programming saying) to promote "composition" over "inheritance". It definitely has its drawbacks too (e.g. more nesting can be a bit harder when editing or looking at JSON), but feels like a cleaner separation to me.

That said, if folks prefer the flatter structure of mixing everything together, we could definitely switch. I just want to be consistent, whatever we decide to do. So for example, a line chart panel definition basically ends up as this right now:

```ts
interface LineChartPanel {
  kind: 'LineChart';

  // These are the common options for all panels
  display: {
    name: string;
    description?: string;
  };

  // These are line chart specific
  options: {
    queries: GraphQueryDefinition[];
    show_legend?: boolean;
  };
}
```

Under a flatter structure, we'd end up with:

```ts
interface LineChartPanel {
  kind: 'LineChart';

  // These are the common options for all panels
  display: {
    name: string;
    description?: string;
  };

  // These are line chart specific
  queries: GraphQueryDefinition[];
  show_legend?: boolean;
}
```

The thing I really like about nesting the plugin-specific stuff under `options` is that it lets us iterate on the common properties without fear of clobbering / conflicting with the plugin-specific ones. But again, totally up for debate what other folks prefer.

**Reply by Nexucis** [reply]
actually I'm not challenging the fact to put the different parameter of the chart in the field `options`. I'm more challenging the choice of the word `options` vs `spec`. `spec` is used already to define the parameter of a resource and I would tend to use it when we want to define a specification of any kind of resource even if it's a nested one such as the chart.

So instead of having 

```typescript
interface LineChartPanel {
  kind: 'LineChart';

  // These are the common options for all panels
  display: {
    name: string;
    description?: string;
  };

  // These are line chart specific
  options: {
    queries: GraphQueryDefinition[];
    show_legend?: boolean;
  };
}
```

we would have 

```typescript
interface LineChartPanel {
  kind: 'LineChart';

  // These are the common options for all panels
  display: {
    name: string;
    description?: string;
  };

  // These are line chart specific
  spec: {
    queries: GraphQueryDefinition[];
    show_legend?: boolean;
  };
}
```

**Comment by Nexucis** [comment]
This is a new draft about how we will define the datasource in Perses. It is following the discussion we had Monday 18th
July.

## How datasource can be defined

Like in the first draft we will have three different ways to define a datasource, depending on the scope we want to give
to the datasource.

### Dashboard level

Like in the first draft, we can extend the dashboard spec to include a list of datasources:

```typescript
interface DashboardSpec {
    // ... existing dashboard spec ...
    datasources: Record<string, DatasourceSpec>;
}
```

Of course the scope of such definition is the dashboard where it is defined. It cannot be used outside on the dashboard.

**Note**: We don’t really have a use case in mind, but as it is not really complicated to have it in the dashboard
specification, we decided to support it.
Once we will have user feedbacks, if they don’t like it, it won’t be that hard to remove as well.

### Project level

In case you would like to share a datasource across different dashboard in the **same** project, you will need to create
an object called Datasource.

```typescript
interface Datasource {
    kind: "Datasource";
    metadata: ProjectMetadata;
    spec: DatasourceSpec;
}
```

#### API definition

##### Get a list of datasources

```bash
GET /api/v1/projects/<project_name>/datasources
```

URL query parameters:

- kind = <string> : should be used to filter the list of datasource with a specific kind
- default = <boolean> : should be used to filter the list of datasource to only have the default one. You should have
  one default datasource per kind
- name = <string> : should be used to filter the list of datasource based on the prefix name.

Example:

The following query should return an empty list or a list containing a single Prometheus datasource.

```bash
GET /api/v1/projects/<project_name>/datasources?kind=Prometheus&default=true
```

##### Get a single datasource

```bash
GET /api/v1/projects/<project_name>/datasources/<datasource_name>
```

##### Create a single datasource

```bash
POST /api/v1/projects/<project_name>/datasources
```

##### Update a single datasource

```bash
PUT /api/v1/projects/<project_name>/datasources/<datasource_name>
```

##### Delete a single datasource

```bash
DELETE /api/v1/projects/<project_name>/datasources/<datasource_name>
```

### Global level

When we talk about scope and user permission in a REST API, the easiest way is to associate one permission per endpoint.
If we want to provide a datasource shared by all projects, then it makes sense to create a different object that is
living outside a project.

That’s why we will have a new resource called `GlobalDatasource`

```typescript
interface GlobalDatasource {
    kind: "GlobalDatasource";
    metadata: Metadata;
    spec: DatasourceSpec;
}
```

Note that the `metadata` is not the same type as in the resource `Datasource`.

#### API definition

##### Get a list of global datasources

```bash
GET /api/v1/globaldatasources
```

URL query parameters:

- kind = <string> : should be used to filter the list of datasource with a specific kind
- default = <boolean> : should be used to filter the list of datasource to only have the default one. You should have
  one default datasource per kind
- name = <string> : should be used to filter the list of datasource based on the prefix name.

Example:

The following query should return an empty list or a list containing a single Prometheus datasource.

```bash
GET /api/v1/globaldatasources?kind=Prometheus&default=true
```

##### Get a single datasource

```bash
GET /api/v1/globaldatasources/<name>
```

##### Create a single datasource

```bash
POST /api/v1/globaldatasources
```

##### Update a single datasource

```bash
PUT /api/v1/globaldatasources/<name>
```

##### Delete a single datasource

```bash
DELETE /api/v1/globaldatasources/<name>
```

### Reason why we don't provide a single object containing a list of datasource

We are wishing to provide a REST API that exposes a way to manage the datasource per project and globally. When we talk
about REST API, it means for every resource provided it comes with the same (at least) 5 different endpoints:

* `GET /<plural resource like dashboardS>` : to be used to return a list of the resource
* `GET /<plural resource>/<resource_name>` : to be used to return a unique resource
* `POST /<plural resource>` : to be used to create a new resource
* `PUT /<plural resource>/<resource_name>` : to be used to update a resource
* `DELETE /<plural resource>/<resource_name>` : to be used to delete a resource

1. If we provide an object like proposed in the first draft

```typescript
// At the Project level
interface ProjectConfigSpec {
    datasources: Record<string, DatasourceDefinition>;
    defaultDatasources: Array<KeyDatasourceSelector>;
}
```

then with the first endpoint that is returning a list of resources, we will have a list of map of datasource.
Not a good thing in this context.

It is understandable using a map will make easier the research, the filtering ...etc. But that's totally something that
be considered. We can add more query parameter to the first endpoint, so we can filter the datasource per kind for
example (like it is proposed above)

2. Providing these endpoints for each resource supported helps to maintain a clean code in the backend as the code won't
   really make a difference between handling a datasource or dashboard. The endpoints will be similar, so the code can
   be a bit more generic.

3. We want to install Perses natively on Kubernetes using CRDs (Custom Resource Documents). Kubernetes will
   automatically expose the different endpoints described above, so we have to respect these standards to be sure we
   won't have any weird behavior when installing Perses on Kubernetes.

## Struct of DatasourceSpec

Like the Panels, we have a common base shared by every kind of datasource:

```typescript
interface DatasourceSpec {
    kind: string; // Like 'Prometheus'
    display?: { // Totally optional
        name: string
        description?: string
    };
    default: boolean // if true, then it's the default datasource
    spec: any // here you will have the specific configuration of the datasource itself
}
```

### Prometheus Datasource

Prometheus as a datasource is basically an HTTP server, so we will likely have just an HTTP config

```typescript
interface PrometheusDatasourceSpec {
    http_config: {
        // url is the url required to contact the datasource
        url: string;
        // The way the UI will contact the datasource. Or through the Backend or directly.
        // By default, access is set with the value 'server'
        access?: 'browser' | 'server';
        //  allowed_endpoints is a list of tuple of http method and http endpoint that will be accessible.
        // These parameters are only used when access is set to 'server'
        allowed_endpoints?: {
            endpoint_pattern: RegExp;
            method: 'POST' | 'PUT' | 'PATCH' | 'GET' | 'DELETE'
        }[];
        // headers can be used to provide additional header that needs to be forwarded when requesting the datasource
        // When defined, it's impossible to set the value of Access with 'browser'
        headers?: Record<string, string>
    };
    // secret contains the name of the resource called secret that holds any password, authentication, certificate, token that should be used to contact the datasource
    secret?: string
    // discovery (for later) will contain the configuration required to contact a remote API that will return the datasource URL.
    discovery?: any
    // TODO @LukeTillman Do you think we could provide a list of variable definitions in the datasource spec ?
    // Like we are making some code injection :D
    variables?: any
}
```

A simple Prometheus datasource would be

```json
{
  "kind": "Datasource",
  "metadata": {
    "name": "PrometheusDemo",
    "project": "perses"
  },
  "spec": {
    "kind": "Prometheus",
    "default": true,
    "spec": {
      "http_config": {
        "url": "https://prometheus.demo.do.prometheus.io",
        "access": "browser"
      }
    }
  }
}
```

A more complex one:

```json
{
  "kind": "Datasource",
  "metadata": {
    "name": "PrometheusDemo",
    "project": "perses"
  },
  "spec": {
    "kind": "Prometheus",
    "default": true,
    "spec": {
      "http_config": {
        "url": "https://prometheus.demo.do.prometheus.io",
        "access": "server",
        "allowed_endpoints": [
          {
            "endpoint_pattern": "/api/v1/labels",
            "method": "POST"
          },
          {
            "endpoint_pattern": "/api/v1/series",
            "method": "POST"
          },
          {
            "endpoint_pattern": "/api/v1/metadata",
            "method": "GET"
          },
          {
            "endpoint_pattern": "/api/v1/query",
            "method": "POST"
          },
          {
            "endpoint_pattern": "/api/v1/query_range",
            "method": "POST"
          },
          {
            "endpoint_pattern": "/api/v1/label/([a-zA-Z0-9_-]+)/values",
            "method": "GET"
          }
        ]
      },
      "secret": "prometheus_secret_config"
    }
  }
}
```

### Selecting / Referencing a Datasource

In the panels, you will be able to select a datasource. Like proposed in the first draft, the selector will be like
this:

```typescript
interface DatasourceSelector {
    kind: string;
    // name is the unique name of the datasource to use
    // If name is omitted, that effectively means "use the default datasource for this kind".
    name?: string;
}
```

Just as a reminder, priority is `"Local datasource in the dashboard" > "Project datasource" > "Global datasource"`.
So if by any chance you a local datasource that is named exactly like a Project datasource, or a Global datasource, we
will consider that the user intentionally wanted to override the upper datasource, so we will use the one with the
smallest scope.

In case we have feedback that ask explicitly to have a way to select precisely what datasource to be used, we will add
another field in the selector like `level` which will indicate at what level the datasource should be retrieved.

> TODO/TO BE DISCUSSED: Selector is at panel level or at query level ?

> TODO/TO BE DISCUSSED: It could be interesting to integrate immediately the field `level` to cover the following
> scenario:
> 1. You created a dashboard where you are selecting a Global Datasource called "Foo".
> 2. After that, someone is creating a new datasource called "Foo" in the same project as the previous dashboard.
> As a result, the scope of the datasource selected change without modifying the dashboard. It can be a good or a bad thing. Somehow I have a feeling it's more a bad thing that a good one.

**Comment by LukeTillman** [comment]
Overall, this is great, thanks for writing it up! The one part I think we dug into a bit in our first meeting (but didn't really talk about much after) was the idea of splitting up things related to the server-proxy into their own resource/config so that we don't have to do anything to prevent potentially sensitive details from leaking into the UI. Do we need to discuss that some more?

I think that might end up looking like:
```ts
interface PrometheusDatasourceSpec {
  // url is the url required to contact the datasource
  url: string;
  // TODO: Maybe some basic auth options at some point as needed from the browser?
}

// Just some pseudo-code (these would probably be multiple different types and probably not in TypeScript at all
// since the UI wouldn't need to know about them unless we want you to be able to edit the settings via the UI)
interface PrometheusDataSourceProxy {
  kind: 'DatasourceProxy';
  spec: {
    kind: 'PrometheusProxy';
    spec: {
      //  allowed_endpoints is a list of tuple of http method and http endpoint that will be accessible.
      // These parameters are only used when access is set to 'server'
      allowed_endpoints?: {
          endpoint_pattern: RegExp;
          method: 'POST' | 'PUT' | 'PATCH' | 'GET' | 'DELETE'
      }[];
      // headers can be used to provide additional header that needs to be forwarded when requesting the datasource
      headers?: Record<string, string>;
      // secret contains the name of the resource called secret that holds any password, authentication, certificate, token that should be used to contact the datasource
      secret?: string
      // discovery (for later) will contain the configuration required to contact a remote API that will return the datasource URL.
      discovery?: any
    }
  }
}
```

Maybe some of those options that are in the proxy are common to multiple datasource types and could be moved up into the DatasourceProxy spec? In a setup like this, though, the front-end would just talk to the proxy like any other Prom server and not need to know the details of the proxy. Also allows the backend to iterate on this feature without breaking the UI's type definitions since the UI wouldn't have any for the proxy-related stuff.

Anyway, something I think to discuss a bit more. Overall though, I think the proposal looks good. Will follow-up on a couple more of the TODOs in there.

**Reply by LukeTillman** [reply]
One other random thought, you could basically generate the Prometheus Datasource config for the UI based on that proxy config. You wouldn't necessarily even need the user to define the separate resource (although maybe you want to for editing/permissions purposes).

**Reply by Nexucis** [reply]
maybe we should discuss it furthermore yeah.

I think we should already considered another kind of datasource to see if our model is extensible enough.

On my side, the fact the secrets are in a different object can prevent to leak any kind of sensitive data.
On the other hand, if we have the variables definitions in the datasource, then I think it makes totally sense to differentiate the way to contact the datasource (so the proxy) and the definition of the datasource itself.

What about something like that:

```typescript
interface PrometheusDatasourceSpec {
    url?: string;
    proxy?: HTTPProxy;
    variables?: any
}

interface basicProxy {
    secret?: string;
    discovery?: any;
}

interface HTTPProxy extends basicProxy {
    kind: "HTTP";

    spec: {
        allowed_endpoints?: {
            endpoint_pattern: RegExp;
            method: 'POST' | 'PUT' | 'PATCH' | 'GET' | 'DELETE'
        }[];
        headers?: Record<string, string>
    };
}

interface SQLDatasourceSpec {
    proxy: SQLProxy;
    variables?: any
}

interface SQLProxy extends basicProxy {
    kind: "SQL";
    spec: {
        url: string
    }
}
```

I just tried to figure out what could be an SQL datasource spec.

I think `discovery` and `secret` will be common to any proxy datasource. I'm wondering if it makes more sense to have it at `spec` level or at the same level than `kind` like I did for the moment.

**Reply by Nexucis** [reply]
I created this [documentation](https://github.com/perses/perses/pull/493) that contains everything we talked about here or during the meeting which is finalising this topic :).

**Comment by LukeTillman** [comment]
Wanted a separate comment/thread for this:

```ts
  // TODO @LukeTillman Do you think we could provide a list of variable definitions in the datasource spec ?
  // Like we are making some code injection :D
  variables?: any
```

Not sure I'm following the idea. Are you thinking variable definitions could go here, associated with the datasource rather than the dashboard? If so, what would happen to the dashboard-level variables field?

**Reply by Nexucis** [reply]
Yeah that's what I was thinking (as maybe a total crazy idea :)).

And in that case, that would mean in the dashboard for the variable you would have to select the datasource, and then you will select the type of the variable, instead of doing the opposite (a.k.a selecting the type of the variable, then potentially select the datasource)

**Reply by Nexucis** [reply]
after a live discussion, we decided it's something we could come back later.
For the moment we think it can increases the complexity of the datasource interface and we don't want to have a dozen of methods to implement when we want to create a new datasource.
